### DISCUSSION TITLE
Time to delay, prevent relay sticking on SetOption80 Shutter mode 2

### DISCUSSION THREAD

**Discussion by steven1180** [arendst/Tasmota#14678] [Q&A] [answered]
Hello, 

**at first....**
I recently came across Tasmota.
I was overwhelmed by how much Tasmota has to offer.
Kudos to the developers.

description:
I created a circuit board myself with a Wemos Mini for two roller shutter.
It is controlled with ShutterMode2. -> First relay for Power, second relay for direction.
For powerrelay I use a solit state IC with zero crossing detection. 
For the directionrelay I use a relay with changeover contacts.
Unfortunately, it happens from time to time that the directional relay sticks.
I'm assuming the capacitor in the roller shutter motor is generating sparks and the relay is switching much slower than the solid state relay. A variant with two normal relays also had the problem that the relays were wearing out and sticking due to the sparks

Now for my request:
Is it possible to add a time or a delay for the direction relay.
When switching on:
first switch on the direction relay, then after delay the solid state powerrelay.
When switching off:
first switch off the solit state and then after delay the directional relay.
here maybe 300ms should be enough.

it would be nice to have such a possibility.

Many Thanks

**Comment by barbudor** [comment]
What kind of delay are you looking at ?
More than a few tenth of milliseconds could be a problem for the code.

What kind of solid state relay you are using.
I think that there are some where the control is active high and others where the control is active low. By combining then you could achieve similar solution using only SSR in a H-bridge using ShutterMode1.
Unfortunately you would need 4 of them for this.

**Comment by steven1180** [comment, marked as answer]
Hi, thanks for the reply.  
The solid state relay is a JC216SC.  
a hard delay is certainly a big problem for the code.  
I expressed myself wrong.  
A delay for the relays could be implemented using a timer.   
I know many situations where motors with capacitors on small relays cause problems.  That's why I chose this JC216SC.  
Regards

**Reply by barbudor** [reply]
Current code doesn't allow easy implementation of an asynchronous delay.
The code is straightforward and synchronous: 
```
                // Code for shutters with circuit safe configuration, switch the direction Relay
                ExecuteCommandPowerShutter(Settings->shutter_startrelay[index] +1, new_shutterdirection == 1 ? 0 : 1, SRC_SHUTTER);
                // power on
                ExecuteCommandPowerShutter(Settings->shutter_startrelay[index], 1, SRC_SHUTTER);
```

**Reply by sfromis** [reply]
If wanting a longer delay in the driver code, you could do the 2nd part a bit later using one of the short-interval calls you can get.

Of course, then it is no longer just a simple change.

**Comment by steven1180** [comment]
Hello, thanks for the idea.  
I'll take a look and try it over the week.  
Maybe it will be included in an update at some point.

Thanks

**Reply by stefanbode** [reply]
Ok, reading again and now I got it:
Your mechanical direction relay is slower then the electrical one. Let me think about this one. The problem is, that a delay would cause some other side-effects on calculating the duration.

This is the answer just for operation delay:
There is another option to get a full stop and delay before changing the direction. There is a hardcoded option:
  delay(MOTOR_STOP_TIME);
This should! stop the motor and all movement for 0.5 seconds before the relay trigger ON again and then may go in the opposite direction. To make this work you have to use the shuttercommands. A direct usage of the relay is forbidden. 
If you like you can compile your own version with MOTOR_STOP_TIME=2000. Then the shutter should wait 2sec before doing anything else.

**Reply by stefanbode** [reply]
I submitted a PR to the driver that should fix your issue. The default wait for the direction relay is now 0.1sec and can be changed by the compiler option. 
#define SHUTTER_RELAY_OPERATION_TIME = 100 // wait for direction relay 0.1sec before power up main relay

**Reply by steven1180** [reply]
can this also be set via the console? like SHUTTER_RELAY_OPERATION_TIME 300?

**Reply by barbudor** [reply]
No, it's a compile time option. Read https://tasmota.github.io/docs/Compile-your-build/

**Reply by stefanbode** [reply]
But it was s now already 0.1 sec. Give it a try

**Reply by steven1180** [reply]
i will try it, thanks

**Reply by sfromis** [reply]
Anyway, you should probably not make `SHUTTER_RELAY_OPERATION_TIME` too much longer, as it is using `delay` which may impede normal Tasmota operation due to everything being blocked for the interval.

**Reply by steven1180** [reply]
Is the reverse also implemented.  When switching off, the ssd relay must go off first and then the direction relay.  This would protect the mechanical relay

**Comment by stefanbode** [comment]
There is a new development version committed just a few seconds ago that should solve your problem. If the default value does not make it just feel free to override the 0.1sec delay.

**Comment by stefanbode** [comment]
Why do you use the expensive JC216SC? ok, quite fancy with the 16A and zero crossing. But what the heck are you moving? I often use the simple HF46F-G (7A). Easy cheap and do the work.

**Comment by steven1180** [comment]
Hello and thanks for the interesting posts.

I now have to translate everything with my somewhat moderate English..... thank google.

I am pleased that this problem now offers a "normal" solution.

I use it to switch capacitor motors.
The contacts of the contactors are always heavily stressed by the capacitors. sticking often occurred.
the motors need about 8A. and to be on the safe side, I used the 16s.
for smaller things is determined also the other suggestion.

I'll try it now and report the result

**Comment by steven1180** [comment]
So. after a long search and read a bit into "visual studio code" to compile it. was i able to successfully put it on my chip.
I was then able to import the configuration file and tested it first with LED`S. you can now see a time interval between the switching actuators. first the click from the relay, and then the LED from the JC216SC which I soldered in front of the control. it looks very nice.
i think this way was important in order to conserve the relays in the future. currently I have
#define SHUTTER_RELAY_OPERATION_TIME = 300
adjusted to be on the safe side. if it doesn't cause any problems, i will keep it or shorten it if necessary until it runs stable.

It would be nice if something were introduced here that works without delay, but according to timer clocks. This could then perhaps be changed via console.

to keep the whole thing free and open, i could even imagine a second timer for switching off

power on:
directional relay on (immediately) and then powerrelay(with time on)

power off:
power relay (immediately) and then direction relay off (with time off)

this would ensure that only one relay switches the current

but so far thank you very much for your help

**Comment by stefanbode** [comment]
I checked the POWER off. For sure there is no delay here either. Anyhow: Is this really required? Because the solid state for sure will go first. But this is the fast relay and there is at least some [ms] before the command comes for the other direction relay. Its line 506ff doing this:
```
      for (int8_t k=0;k<2;k++) {
        if ((1 << (Settings->shutter_startrelay[i]+k-1)) & TasmotaGlobal.power) {
          ExecuteCommandPowerShutter(Settings->shutter_startrelay[i]+k, 0, SRC_SHUTTER);
        }
      }
```

**Reply by stefanbode** [reply]
Regarding the "delay". Yes this is the ONLY option to ensure I get control back and continue. Also it prevent side effects. Therefore this value should a low as possible. The 300ms should be no big deal here.

**Comment by steven1180** [comment]
Yes, you are right, I hope and also think that the solit state will also be faster than the relay when it is switched off.

due to construction work in the house, the hardware test will be delayed a little. but I'm confident.

Thanks very much
